### PR TITLE
REMOVE tslint, and use eslint for everything

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: TSlint will get replaced with ESlint at some point (see [this article](https://medium.com/palantir/tslint-in-2019-1a144c2317a9) for explanations on why), having 2 linters is annoying:

- more processes to run
- more caching to configure
- more editor config

## What I did
- CHANGE the Eslint config so it can parse .ts,.tsx,.mjs files
- REMOVE tslint
- now that eslint rules are having an effect on typescript code, I did a ton of fixes
- had to do some additional fixes to avoid using ts-loader, related: https://github.com/storybooks/storybook/pull/6566

### CODE REVIEW

File: addons/a11y/src/components/Report/HighlightToggle.tsx [hunk 1]
```diff
@@ -99,34 +103,55 @@ class HighlightToggle extends Component<ToggleProps> {
...
 
   componentDidUpdate(prevProps: Readonly<ToggleProps>): void {
+    const { indeterminate } = this.props;
     if (this.checkBoxRef.current) {
-      this.checkBoxRef.current.indeterminate = this.props.indeterminate;
+      this.checkBoxRef.current.indeterminate = indeterminate;
     }
   }
 
+  onToggle = (): void => {
```
Comment (shilman) [inline addons/a11y/src/components/Report/HighlightToggle.tsx]: What are these changes?

Comment (ndelangen) [reply addons/a11y/src/components/Report/HighlightToggle.tsx]: Linting fixes, typescript files were missing a ton of useful eslint rules!

Now all eslint rules apply to typescript code as well 🎉

File: examples/official-storybook/webpack.config.js [hunk 2]
```diff
@@ -5,7 +5,40 @@ module.exports = async ({ config }) => ({
...
-      ...config.module.rules,
+      ...config.module.rules.slice(1),
+      {
+        test: /\.(mjs|jsx?|tsx?)$/,
+        use: [
+          {
+            loader: 'babel-loader',
+            options: {
+              cacheDirectory:
+                '/Users/dev/Projects/GitHub/storybook/core/examples/official-storybook/node_modules/.cache/storybook',
```
Comment (shilman) [inline examples/official-storybook/webpack.config.js]: ?!!!

**Review by shilman — CHANGES_REQUESTED** [review]
Love it! One weird configuration change above, otherwise looks fantastic 🎉

File: addons/a11y/src/components/ColorBlindness.tsx [hunk 3]
```diff
@@ -63,7 +63,8 @@ export class ColorBlindness extends Component<ColorBlindnessProps, ColorBlindnes
   };
 
   onVisibilityChange = (s: boolean) => {
-    if (this.state.expanded !== s) {
+    const { expanded } = this.state;
+    if (expanded !== s) {
```
Comment (gaetanmaisse) [inline addons/a11y/src/components/ColorBlindness.tsx]: Which rule causes this? [destructuring-assignment](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/destructuring-assignment.md) ?

File: addons/contexts/src/shared/types.d.ts [hunk 4]
```diff
@@ -21,11 +22,11 @@ export declare interface AddonSetting {
   icon?: ComponentProps<typeof Icons>['icon'] | '';
   title: string;
   components?: unknown[];
-  params?: Array<{
+  params?: {
     name: string;
     props: GenericProp;
     default?: boolean;
-  }>;
+  }[];
```
Comment (gaetanmaisse) [inline addons/contexts/src/shared/types.d.ts]: Not sure about this one, I think `Array<>` is easier to read. What other devs are thinking about it?

File: examples/cra-ts-kitchen-sink/src/react-app-env.d.ts [hunk 5]
```diff
@@ -1 +1 @@
-/// <reference types="react-scripts" />
+// / <reference types="react-scripts" />
```
Comment (gaetanmaisse) [inline examples/cra-ts-kitchen-sink/src/react-app-env.d.ts]: ⚠️ Nop, this is not a comment but a [triple slash directive](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html).

File: examples/official-storybook/webpack.config.js (hunk 2 above)
Comment (gaetanmaisse) [reply examples/official-storybook/webpack.config.js]: Fixed in next commit :youhou:

File: addons/a11y/src/components/ColorBlindness.tsx (hunk 3 above)
Comment (ndelangen) [reply addons/a11y/src/components/ColorBlindness.tsx]: yeah, something react destructuring props & state

File: addons/contexts/src/shared/types.d.ts (hunk 4 above)
Comment (ndelangen) [reply addons/contexts/src/shared/types.d.ts]: I agree personally on that I find `Array<X>` to be easier to read then `X[]`, but this is a recommended rule from the typescript folk

File: addons/a11y/src/components/ColorBlindness.tsx (hunk 3 above)
Comment (ndelangen) [reply addons/a11y/src/components/ColorBlindness.tsx]: I agree with this rule mostly, though I can be a bit annoying if you only have 1 prop or 1 state property.

**Review by gaetanmaisse — APPROVED** [review]

**Review by shilman — APPROVED** [review]
Beautiful! 👌

File: .eslintrc.js [hunk 6]
```diff
@@ -94,13 +113,21 @@ module.exports = {
       error,
       { allow: ['__STORYBOOK_CLIENT_API__', '__STORYBOOK_ADDONS_CHANNEL__'] },
     ],
+    '@typescript-eslint/no-var-requires': ignore,
+    '@typescript-eslint/camelcase': ignore,
+    '@typescript-eslint/no-unused-vars': ignore,
```
Comment (Hypnosphi) [inline .eslintrc.js]: Why?

File: examples/official-storybook/webpack.config.js [hunk 7]
```diff
@@ -5,7 +5,39 @@ module.exports = async ({ config }) => ({
...
+        test: /\.(mjs|jsx?|tsx?)$/,
+        use: [
+          {
+            loader: 'babel-loader',
+            options: {
+              cacheDirectory: `.cache/storybook`,
+              presets: [
+                ['@babel/preset-env', { shippedProposals: true, useBuiltIns: 'usage', corejs: 2 }],
+                '@babel/preset-typescript',
+                ['babel-preset-minify', { builtIns: false, mangle: false }],
```
Comment (Hypnosphi) [inline examples/official-storybook/webpack.config.js]: @ndelangen shouldn't we apply this one only in prod mode?

Comment (ndelangen) [reply examples/official-storybook/webpack.config.js]: yeah

**Pull request merged**